relayer/workers/parachaincommitmentrelayer/beefy-listener.go

```diff
@@ -156,7 +164,7 @@ func (li *BeefyListener) processBeefyLightClientEvents(ctx context.Context, even
 		// the first parachain block that has not yet been fully processed on ethereum,
 		// and then package and relay all newer heads/commitments together with their corresponding leaf
 		mmrProof := li.relaychainConn.GetMMRLeafForBlock(uint64(relayChainBlockNumber-1), blockHash)
-		allParaHeads, ourParaHead := li.GetAllParaheads(blockHash, OUR_PARACHAIN_ID)
+		allParaHeads, ourParaHead := li.relaychainConn.GetAllParaheadsWithOwn(blockHash, OUR_PARACHAIN_ID)
```

Comment:
There's still the fundamental issue in that we're only analysing a single MMR leaf in between justifications, meaning we're skipping over para heads that could contain commitments.

Its not clear to me how this would be mitigated by fixing catchup. How would the worker know when to restart and catchup? Even if we overcome that, it just doesn't seem right for the worker to crash and restart so very often. This perhaps points to a need to redesign the listener again, and decoupling commitment listening from justification listening.

Anyhow, the situation with Polkadot v0.9.4 is that justifications get emitted periodically (not sure if its regular or not). For example, suppose justifications get emitted at relay chain block 0, 10, 20. Furthermore, our parachain produces new blocks on the same interval as the relay chain. This code, which seems limited to discovering a single para head in between justifications, would only discover 3 of our parachain heads, and ignore 17 other heads. So we're potentially ignoring 17 commitments.

So it seems we possibly do need to be analyzing `Paras.Heads` over a range of relay chain blocks `[M, N]`, where M and N are the relay chain blocks for two successive justifications?

Reply:
Good observation – the original architecture foresaw to have them decoupled. The idea was to listen to any new commitments and put them in a queue. Whenever a new beefy MMR root is observed, check the queue for commitments that can be proven now, and submit them one by one.

Reply:
> So it seems we possibly do need to be analyzing `Paras.Heads` over a range of relay chain blocks `[M, N]`, where M and N are the relay chain blocks for two successive justifications?

@vgeddes this PR is still WIP, sorry - should have mentioned that the review is a bit premature. the intention is to implement exactly this and to use it for both the catchup and for the normal relaying flow